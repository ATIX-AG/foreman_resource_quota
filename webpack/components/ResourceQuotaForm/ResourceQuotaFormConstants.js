/* Resource identifier */
export const RESOURCE_IDENTIFIER_ID = 'id';
export const RESOURCE_IDENTIFIER_NAME = 'name';
export const RESOURCE_IDENTIFIER_DESCRIPTION = 'description';
export const RESOURCE_IDENTIFIER_CPU = 'cpu_cores';
export const RESOURCE_IDENTIFIER_MEMORY = 'memory_mb';
export const RESOURCE_IDENTIFIER_DISK = 'disk_gb';
export const RESOURCE_IDENTIFIER_STATUS_NUM_HOSTS = 'number_of_hosts';
export const RESOURCE_IDENTIFIER_STATUS_NUM_USERS = 'number_of_users';
export const RESOURCE_IDENTIFIER_STATUS_NUM_USERGROUPS = 'number_of_usergroups';
export const RESOURCE_IDENTIFIER_STATUS_UTILIZATION = 'utilization';
export const RESOURCE_IDENTIFIER_STATUS_MISSING_HOSTS = 'missing_hosts';

/* Resource names */
export const RESOURCE_NAME_CPU = 'CPU cores';
export const RESOURCE_NAME_MEMORY = 'Memory';
export const RESOURCE_NAME_DISK = 'Disk space';

/* Resource units (order the units with increasing factor!) */
export const RESOURCE_UNIT_CPU = [{ symbol: 'cores', factor: 1 }];
export const RESOURCE_UNIT_MEMORY = [
  { symbol: 'MB', factor: 1 },
  { symbol: 'GB', factor: 1024 },
  { symbol: 'TB', factor: 1024 * 1024 },
];
export const RESOURCE_UNIT_DISK = [
  { symbol: 'GB', factor: 1 },
  { symbol: 'TB', factor: 1024 },
  { symbol: 'PB', factor: 1024 * 1024 },
];

/* Resource value bounds */
export const RESOURCE_VALUE_MIN_CPU = 0;
export const RESOURCE_VALUE_MAX_CPU = 1999999999;
export const RESOURCE_VALUE_MIN_MEMORY = 0;
export const RESOURCE_VALUE_MAX_MEMORY = 1999999999;
export const RESOURCE_VALUE_MIN_DISK = 0;
export const RESOURCE_VALUE_MAX_DISK = 1999999999;

/* Map attributes to given resource identifier (name, unit, minValue, maxValue) */
export const resourceAttributesByIdentifier = identifier => {
  switch (identifier) {
    case RESOURCE_IDENTIFIER_CPU:
      return {
        name: RESOURCE_NAME_CPU,
        unit: RESOURCE_UNIT_CPU,
        minValue: RESOURCE_VALUE_MIN_CPU,
        maxValue: RESOURCE_VALUE_MAX_CPU,
      };
    case RESOURCE_IDENTIFIER_MEMORY:
      return {
        name: RESOURCE_NAME_MEMORY,
        unit: RESOURCE_UNIT_MEMORY,
        minValue: RESOURCE_VALUE_MIN_MEMORY,
        maxValue: RESOURCE_VALUE_MAX_MEMORY,
      };
    case RESOURCE_IDENTIFIER_DISK:
      return {
        name: RESOURCE_NAME_DISK,
        unit: RESOURCE_UNIT_DISK,
        minValue: RESOURCE_VALUE_MIN_DISK,
        maxValue: RESOURCE_VALUE_MAX_DISK,
      };
    default:
      return null;
  }
};

/* HTML constants */
export const MODAL_ID_CREATE_RESOURCE_QUOTA = `foreman-resource-quota-create-modal`;
export const MODAL_ID_UPDATE_RESOURCE_QUOTA = `foreman-resource-quota-create-modal`;
