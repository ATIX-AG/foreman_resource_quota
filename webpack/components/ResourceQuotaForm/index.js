import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import { Gallery, GalleryItem } from '@patternfly/react-core';

import SkeletonLoader from 'foremanReact/components/common/SkeletonLoader';
import { STATUS } from 'foremanReact/constants';

import Properties from './components/Properties/';
import Resources from './components/Resources';
import Submit from './components/Submit';
import QuotaState from './components/QuotaState';
import {
  RESOURCE_IDENTIFIER_ID,
  RESOURCE_IDENTIFIER_NAME,
  RESOURCE_IDENTIFIER_DESCRIPTION,
  RESOURCE_IDENTIFIER_CPU,
  RESOURCE_IDENTIFIER_MEMORY,
  RESOURCE_IDENTIFIER_DISK,
  RESOURCE_IDENTIFIER_STATUS_NUM_HOSTS,
  RESOURCE_IDENTIFIER_STATUS_NUM_USERS,
  RESOURCE_IDENTIFIER_STATUS_NUM_USERGROUPS,
  RESOURCE_IDENTIFIER_STATUS_MISSING_HOSTS,
  RESOURCE_IDENTIFIER_STATUS_UTILIZATION,
} from './ResourceQuotaFormConstants';

const ResourceQuotaForm = ({
  isNewQuota,
  initialProperties,
  initialStatus,
  onSubmit,
  quotaChangesCallback,
}) => {
  const modelState = QuotaState({
    initialProperties,
    initialStatus,
    changesCallback: quotaChangesCallback,
  });

  const [isLoading, setIsLoading] = useState(!isNewQuota);
  const [fetchedOnce, setFetchedOnce] = useState(false);

  const fetchCallback = useCallback(() => {
    setIsLoading(false);
  }, []);

  const fetchQuotaOnce = useCallback(() => {
    if (!isNewQuota && !fetchedOnce) {
      modelState.onFetch(fetchCallback);
      setFetchedOnce(true);
    }
  }, [modelState, isNewQuota, fetchedOnce, fetchCallback]);

  useEffect(() => {
    fetchQuotaOnce(); // this will be executed only on the first render (valid is ignored)
  }, [fetchQuotaOnce]);

  return (
    <Gallery maxWidths={{ md: '800px' }}>
      <Gallery hasGutter>
        <GalleryItem key="edit-resource-quota-properties-item">
          <SkeletonLoader
            skeletonProps={{ width: 300 }}
            status={isNewQuota || !isLoading ? STATUS.RESOLVED : STATUS.PENDING}
          >
            {(!isLoading || isNewQuota) && (
              <Properties
                isNewQuota={isNewQuota}
                initialName={modelState.getQuotaProperties(
                  RESOURCE_IDENTIFIER_NAME
                )}
                initialDescription={modelState.getQuotaProperties(
                  RESOURCE_IDENTIFIER_DESCRIPTION
                )}
                initialStatus={modelState.getQuotaStatus()}
                handleInputValidation={modelState.handleInputValidation}
                onChange={modelState.onChange}
                onApply={modelState.onApply}
                onFetch={modelState.onFetchUtilization}
              />
            )}
          </SkeletonLoader>
        </GalleryItem>
        <GalleryItem key="edit-resource-quota-resources-item">
          <SkeletonLoader
            skeletonProps={{ width: 400 }}
            status={isNewQuota || !isLoading ? STATUS.RESOLVED : STATUS.PENDING}
          >
            {(!isLoading || isNewQuota) && (
              <Resources
                isNewQuota={isNewQuota}
                initialProperties={modelState.getQuotaProperties()}
                initialStatus={modelState.getQuotaStatus(
                  RESOURCE_IDENTIFIER_STATUS_UTILIZATION
                )}
                handleInputValidation={modelState.handleInputValidation}
                onChange={modelState.onChange}
                onApply={modelState.onApply}
              />
            )}
          </SkeletonLoader>
        </GalleryItem>
        {isNewQuota && (
          <GalleryItem key="edit-resource-quota-submit-item">
            <Submit
              isValid={modelState.getIsValid()}
              onCreate={modelState.onCreate}
              onSubmit={onSubmit}
            />
          </GalleryItem>
        )}
      </Gallery>
    </Gallery>
  );
};

ResourceQuotaForm.defaultProps = {
  onSubmit: null,
  quotaChangesCallback: null,
  initialProperties: {
    [RESOURCE_IDENTIFIER_NAME]: '',
    [RESOURCE_IDENTIFIER_DESCRIPTION]: '',
    [RESOURCE_IDENTIFIER_CPU]: null,
    [RESOURCE_IDENTIFIER_MEMORY]: null,
    [RESOURCE_IDENTIFIER_DISK]: null,
  },
  initialStatus: {
    [RESOURCE_IDENTIFIER_STATUS_NUM_HOSTS]: null,
    [RESOURCE_IDENTIFIER_STATUS_NUM_USERS]: null,
    [RESOURCE_IDENTIFIER_STATUS_NUM_USERGROUPS]: null,
    [RESOURCE_IDENTIFIER_STATUS_MISSING_HOSTS]: null,
    [RESOURCE_IDENTIFIER_STATUS_UTILIZATION]: {
      [RESOURCE_IDENTIFIER_CPU]: null,
      [RESOURCE_IDENTIFIER_MEMORY]: null,
      [RESOURCE_IDENTIFIER_DISK]: null,
    },
  },
};

ResourceQuotaForm.propTypes = {
  isNewQuota: PropTypes.bool.isRequired,
  onSubmit: PropTypes.func,
  quotaChangesCallback: PropTypes.func,
  initialProperties: PropTypes.shape({
    [RESOURCE_IDENTIFIER_ID]: PropTypes.number,
    [RESOURCE_IDENTIFIER_NAME]: PropTypes.string,
    [RESOURCE_IDENTIFIER_DESCRIPTION]: PropTypes.string,
    [RESOURCE_IDENTIFIER_CPU]: PropTypes.number,
    [RESOURCE_IDENTIFIER_MEMORY]: PropTypes.number,
    [RESOURCE_IDENTIFIER_DISK]: PropTypes.number,
  }),
  initialStatus: PropTypes.shape({
    [RESOURCE_IDENTIFIER_STATUS_NUM_HOSTS]: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.oneOf([null]),
    ]),
    [RESOURCE_IDENTIFIER_STATUS_NUM_USERS]: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.oneOf([null]),
    ]),
    [RESOURCE_IDENTIFIER_STATUS_NUM_USERGROUPS]: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.oneOf([null]),
    ]),
    [RESOURCE_IDENTIFIER_STATUS_MISSING_HOSTS]: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.oneOf([null]),
    ]),
    [RESOURCE_IDENTIFIER_STATUS_UTILIZATION]: PropTypes.shape({
      [RESOURCE_IDENTIFIER_CPU]: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.oneOf([null]),
      ]),
      [RESOURCE_IDENTIFIER_MEMORY]: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.oneOf([null]),
      ]),
      [RESOURCE_IDENTIFIER_DISK]: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.oneOf([null]),
      ]),
    }),
  }),
};

export default ResourceQuotaForm;
