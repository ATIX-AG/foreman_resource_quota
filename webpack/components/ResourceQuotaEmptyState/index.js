import React, { useState } from 'react';
import { Button, Modal, ModalVariant } from '@patternfly/react-core';

import { translate as __ } from 'foremanReact/common/I18n';
import { getDocsURL } from 'foremanReact/common/helpers';
import EmptyStatePattern from 'foremanReact/components/common/EmptyState/EmptyStatePattern';

import ResourceQuotaForm from '../ResourceQuotaForm';
import { MODAL_ID_CREATE_RESOURCE_QUOTA } from '../ResourceQuotaForm/ResourceQuotaFormConstants';

const ResourceQuotaEmptyState = () => {
  const [isOpen, setIsOpen] = useState(false);

  const onSubmitSuccessCallback = success => {
    if (success) {
      setIsOpen(false);
      window.location.reload();
    }
  };

  const ActionButton = (
    <Button
      id="foreman-resource-quota-welcome-create-modal-button"
      ouiaId="foreman-resource-quota-welcome-create-modal-button"
      variant="primary"
      onClick={() => {
        setIsOpen(true);
      }}
    >
      {__('Create resource quota')}
    </Button>
  );

  const description = (
    <span>
      {__(
        'Resource Quotas help admins to manage resources including CPUs, memory, and disk space among users or user groups.'
      )}
      <br />
      {__(
        'Define a Resource Quota here and apply it to users to guarantee a fair share of your resources.'
      )}
      <br />
    </span>
  );
  const documentation = {
    url: getDocsURL('Administering_Project', 'limiting-host-resources'),
  };

  return (
    <div>
      <EmptyStatePattern
        icon="pficon pficon-cluster"
        iconType="pf"
        header={__('Resource Quotas')}
        description={description}
        action={ActionButton}
        documentation={documentation}
      />
      <Modal
        ouiaId={MODAL_ID_CREATE_RESOURCE_QUOTA}
        title={__('Create resource quota')}
        variant={ModalVariant.small}
        isOpen={isOpen}
        onClose={() => {
          setIsOpen(false);
        }}
        appendTo={document.body}
      >
        <ResourceQuotaForm isNewQuota onSubmit={onSubmitSuccessCallback} />
      </Modal>
    </div>
  );
};

export default ResourceQuotaEmptyState;
