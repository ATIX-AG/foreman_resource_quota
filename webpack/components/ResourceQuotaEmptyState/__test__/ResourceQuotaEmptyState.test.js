import React from 'react';
import '@testing-library/jest-dom';

import { mount, testComponentSnapshotsWithFixtures } from '@theforeman/test';
// Notice: (not) importing Modal affects the snapshot test since it fills
// the components data dynamically in snapshots as soon as it can find the component.
import { Modal } from '@patternfly/react-core';

import { withMockedProvider, withRedux } from '../../../test_helper';
import ResourceQuotaForm from '../../ResourceQuotaForm';
import ResourceQuotaEmptyState from '../index';

const TestComponent = withRedux(withMockedProvider(ResourceQuotaEmptyState));

describe('ResourceQuotaEmptyState', () => {
  testComponentSnapshotsWithFixtures(ResourceQuotaEmptyState, {
    'should render': {}, // component has no props
  });

  test('opens the modal on clicking "Create resource quota" button', () => {
    const wrapper = mount(<TestComponent />);

    expect(wrapper.find(Modal).prop('isOpen')).toBe(false); // check we provide the correct input to Modal
    expect(wrapper.find(ResourceQuotaForm).exists()).toBe(false);

    wrapper
      .find('button')
      .filterWhere(button => button.text() === 'Create resource quota')
      .simulate('click');
    wrapper.update();

    expect(wrapper.find(Modal).prop('isOpen')).toBe(true);
    expect(wrapper.find(ResourceQuotaForm).exists()).toBe(true);
  });
});
