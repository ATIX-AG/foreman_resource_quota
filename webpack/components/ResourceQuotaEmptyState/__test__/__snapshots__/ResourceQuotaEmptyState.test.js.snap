// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceQuotaEmptyState should render 1`] = `
<div>
  <EmptyStatePattern
    action={
      <Button
        id="foreman-resource-quota-welcome-create-modal-button"
        onClick={[Function]}
        ouiaId="foreman-resource-quota-welcome-create-modal-button"
        variant="primary"
      >
        Create resource quota
      </Button>
    }
    description={
      <span>
        Resource Quotas help admins to manage resources including CPUs, memory, and disk space among users or user groups.
        <br />
        Define a Resource Quota here and apply it to users to guarantee a fair share of your resources.
        <br />
      </span>
    }
    documentation={
      Object {
        "url": "/links/docs/Administering_Project?chapter=limiting-host-resources",
      }
    }
    header="Resource Quotas"
    icon="pficon pficon-cluster"
    iconType="pf"
    secondaryActions={Array []}
  />
  <Modal
    actions={Array []}
    appendTo={<body />}
    aria-describedby=""
    aria-label=""
    aria-labelledby=""
    className=""
    hasNoBodyWrapper={false}
    isOpen={false}
    onClose={[Function]}
    ouiaId="foreman-resource-quota-create-modal"
    ouiaSafe={true}
    position="default"
    showClose={true}
    title="Create resource quota"
    titleIconVariant={null}
    titleLabel=""
    variant="small"
  >
    <ResourceQuotaForm
      initialProperties={
        Object {
          "cpu_cores": null,
          "description": "",
          "disk_gb": null,
          "memory_mb": null,
          "name": "",
          "unassigned": false,
        }
      }
      initialStatus={
        Object {
          "missing_hosts": null,
          "number_of_hosts": null,
          "number_of_usergroups": null,
          "number_of_users": null,
          "utilization": Object {
            "cpu_cores": null,
            "disk_gb": null,
            "memory_mb": null,
          },
        }
      }
      isNewQuota={true}
      onSubmit={[Function]}
      quotaChangesCallback={null}
      showAssignmentWarning={false}
    />
  </Modal>
</div>
`;
